:books: translate chapter 8. Hints

takaidohigasi opened the pull request:
## 原文

http://www.unofficialmysqlguide.com/hints.html

## 備考

BKA, BNL, MRRなどについては公式マニュアルの表記に準拠

https://dev.mysql.com/doc/refman/5.6/ja/bnl-bka-optimization.html
https://dev.mysql.com/doc/refman/5.6/ja/mrr-optimization.html
https://dev.mysql.com/doc/refman/5.6/ja/index-condition-pushdown-optimization.html
https://dev.mysql.com/doc/refman/5.6/ja/range-optimization.html

* ICP => インデックスコンディションプッシュダウン (※日本語ひらき)
* MRR => Multi-Range Read
* BKA => Batched Key Access 
* BNL => Block Nested Loop 
* range optimization => rangeの最適化

takaidohigasi committed:
:+1: rename post file to unify file name to the original site

takaidohigasi committed:
:+1: improve word order for chapter 3 b+tree

takaidohigasi committed:
:books: translate chapter 8. Hints

takaidohigasi commented:
本人チェックOK

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+---
+title: ヒント句 
```
ヒント句ってOracle用語でしょうか？個人的な好みかもしれませんが、インデックスヒント、あるいはヒントの方がしっくりきます([MySQL 5.6のマニュアル](https://dev.mysql.com/doc/refman/5.6/ja/index-hints.html)にもインデックスヒント、ヒントは出てくるがヒント句はでてこない)。

takaidohigasi replied:
Oracle, PostgreSQLでもいずれも利用する気がしたので確認しました
総称としては「ヒント」が正しそうですのでこれに修正します。

Oracle
https://docs.oracle.com/cd/E49329_01/server.121/b71278/sql_elements006.htm#i35922

PostgreSQL
※最も有名と思われるextention
http://pghintplan.osdn.jp/pg_hint_plan-ja.html

> ヒント句	実行計画を制御するための情報です。
> ヒント	実行計画を制御したいクエリに適用するヒント句を列挙したものです。

ヒントとヒント句は違う。

インデックスヒントはヒントの一部分で(SEMIJOINなどのヒントやQB_NAMEはこれには含まれない)、インデックスに対する機能を担うヒントの総称のようです。

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+1. **デバッグにて** `EXPLAIN`は利用できたメタデータに基づいて決定された結果を表示します。ヒント句を追加することによって、他の実行計画を強制的に実行させ、実際の実行時間を比較することができます。
+2. **商用環境にて** デバッグの最中に代わりとなる実行計画が十分に速いと気づいたのであれば、実行を高速化するためにヒント句を加えるとよいでしょう。
+
+私はデバッグではヒント句をよく利用しますが、商用環境に導入するときにはいつも気をつけています。ヒント句はある時点で選択された特定の実行計画に固定するものであるため、データの分布やインデックスが時とともに変化した場合に維持管理上の重荷になりえます。
```
7章の訳で指摘したように、productionを本番と訳すならこれも合わせる必要がありますね。

takaidohigasi replied:
ありがとうございます、修正いたします！

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+
+私はデバッグではヒント句をよく利用しますが、商用環境に導入するときにはいつも気をつけています。ヒント句はある時点で選択された特定の実行計画に固定するものであるため、データの分布やインデックスが時とともに変化した場合に維持管理上の重荷になりえます。
+
+ベストプラクティスとしては、MySQLのメジャーバージョンアップグレードをした後(例えば、MySQL 9.0にアップグレードするときに)には毎回ヒント句を見直すと良いでしょう[^1]。新しい最適化が導入されると、それに伴いたくさんのヒントを無効化できることがお分かりになることでしょう。
```
英語のupgradeは日本語ではバージョンアップと言ってしまうことが多いと思うので、ここでは「MySQLのメジャーバージョンアップ」でもいいと思います。

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+MySQL 8.0ではMySQL 5.7で導入された新しいコメント形式のヒント句を拡張し、テーブルの結合順序を(`STRAIGHT_JOIN`と同様に)制御できるようになりました。私は次の3つの理由から、古いSQLの文法拡張したヒント句よりこの新しいヒント句の方が好みです。
+
+1. どのように実行するかを教えるといったといったことは、SQLの宣言型の本質的なものと分離したほうが読みやすいし書きやすい
+2. 「ヒント」であって命令型でないというとう明確なセマンティクスであること。すなわち、ヒントが実行できなかった場合、ステートメントは警告を発生させるもののエラーは発生させないということになります。これは`FORCE INDEX`などのヒントとは対照的で、この場合はインデックスが存在しないとエラーを発生させます。これは、例えばOracleデータベースのヒント句と似た挙動です。
```
命令形でないというとう → 命令形でないという

takaidohigasi replied:
ありがとうございます！

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+MySQL 8.0ではMySQL 5.7で導入された新しいコメント形式のヒント句を拡張し、テーブルの結合順序を(`STRAIGHT_JOIN`と同様に)制御できるようになりました。私は次の3つの理由から、古いSQLの文法拡張したヒント句よりこの新しいヒント句の方が好みです。
+
+1. どのように実行するかを教えるといったといったことは、SQLの宣言型の本質的なものと分離したほうが読みやすいし書きやすい
+2. 「ヒント」であって命令型でないというとう明確なセマンティクスであること。すなわち、ヒントが実行できなかった場合、ステートメントは警告を発生させるもののエラーは発生させないということになります。これは`FORCE INDEX`などのヒントとは対照的で、この場合はインデックスが存在しないとエラーを発生させます。これは、例えばOracleデータベースのヒント句と似た挙動です。
```
semanticsは単に「意味」の方が分かりやすいかもしれません。

takaidohigasi replied:
意味がわかりづらいのでちょっと見直してみました。

```
「ヒント」であって命令でないことを構造的に明確に示していること。
```

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+| SEMIJOIN, NO\_SEMIJOIN | サブクエリーを実行する際の準結合戦略(DUPSWEEDOUT, FIRSTMATCH, LOOSESCAN, MATERIALIZATIONなど) |
+| SUBQUERY | SEMIJOIN/NO_SEMIJOINと類似しており、IN-to-EXISTSなども含めたサブクエリー実行の際の戦略を制御できます |
+
+例6では、あるテーブルに対してrange最適化が無効化されています。これによりとても良い方法であるにも関わらず、p(population)インデックスが利用されません。population > 10億であるものはテーブル中に2行しかありません。
```
訳では英語の順と逆になっているのは日本語として自然なのでいいと思いますが、「これによりとても良い方法であるにも関わらず」よりも、「p(population)インデックスを使用した方がずっと良いにも関わらず、無視されてしまいます」とした方が「とても良い方法」とは何かが明確になって良いと思います。

takaidohigasi replied:
ありがとうございます、コメントいただいたほうの役のほうが分かりやすいので修正します。
日本語力をあげなくては...

takaidohigasi committed:
:bug: fix word 「ヒント句」 to 「ヒント」

takaidohigasi committed:
:books: improve chapter 8 translations

- fix word 「商用」 to 「本番」
- 「メジャーバージョンアップグレード」to 「メジャーバージョンアップ」
- fix typo 「命令形でないというとう」
- improve translation for semantics
- improve traslated japanese word order

takaidohigasi committed:
:books: fix word from 「コスト単位」to 「コストユニット」

takaidohigasi commented:
ヒント句 → ヒントの修正については量が多いためコミットを分けました
https://github.com/yakst/unofficialmysqlguide-ja/pull/38/commits/dd87bce4ee642a15647b1fe360d0cc9a2e1fe4f6

その他の修正
https://github.com/yakst/unofficialmysqlguide-ja/pull/38/commits/58517ce8f4a18670621fff143c42f4adb36a0a5b

「コスト単位」 > 「コストユニット」の修正もついでにコミット積みました
https://github.com/yakst/unofficialmysqlguide-ja/pull/38/commits/9db20c8e86f974a33cce497652b6c794ced7e8c5
※IDEで全検索置換

doublemarket commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+
+MySQL 8.0ではMySQL 5.7で導入された新しいコメント形式のヒントを拡張し、テーブルの結合順序を(`STRAIGHT_JOIN`と同様に)制御できるようになりました。私は次の3つの理由から、古いSQLの文法拡張したヒントよりこの新しいヒントの方が好みです。
+
+1. どのように実行するかを教えるといったといったことは、SQLの宣言型の本質的なものと分離したほうが読みやすいし書きやすい
```
といった、が2回になっていますね。

takaidohigasi replied:
ありがとうございます！読みかえしても気がつけませんでしたっ

doublemarket reviewed (approved):
1点だけ指摘しましたがそれ以外はLGTM!

takaidohigasi commented on _posts/2017-03-14-hints.md:
```diff
@@ -0,0 +1,170 @@
+
+MySQL 8.0ではMySQL 5.7で導入された新しいコメント形式のヒントを拡張し、テーブルの結合順序を(`STRAIGHT_JOIN`と同様に)制御できるようになりました。私は次の3つの理由から、古いSQLの文法拡張したヒントよりこの新しいヒントの方が好みです。
+
+1. 「どのように実行するかを教える」といったことは、SQLの宣言型の本質的なものと分離したほうが読みやすいし書きやすい
```
「といった」の重複を修正しました
